First seeding prototype based on in functions

Lexolef opened the pull request:
It is not yet complete but it gets a 96% on the in testcases.
This version still struggles with modified parameters and values like .lower, but Sharui and I are working on that.
I added a few suggestions in the descriptions and comments. So your opinion would be nice.
But this PR is mainly for your suggestions and maybe advise on how we should proceed.

Have fun :D

Lexolef committed:
First seeding strategy

Lexolef committed:
Refactored code (untested)

Lexolef committed:
Small refactoring fix

Lexolef committed:
Generalized functions for future use

Lexolef committed:
First version handling and, in statements

Lexolef committed:
Implemented and statement recognition

Lexolef committed:
Added comments and removed unnecessary statements

Lexolef committed:
Merge branch 'main' of https://github.com/Priapos1004/penguin_and_fuzzingbook into seeding_prototype

K92hoang committed:
Fixed function signature in str_comp tests

K92hoang committed:
Fixed function signature in in_func tests

K92hoang committed:
Added len_func tests

K92hoang committed:
Fixed three tests

Lexolef committed:
Fixed and accumulation

Lexolef committed:
Added multiple if for one parameter

Lexolef committed:
Merge branch 'len-func-tests' of https://github.com/Priapos1004/penguin_and_fuzzingbook into seeding_prototype

Lexolef committed:
Added or coverage

Lexolef committed:
Removed unnecessary parameter

Lexolef committed:
Removed unnecessary return

Lexolef committed:
corrected ruff format

Lexolef committed:
Final comments and clean up checks

sharuiY committed:
Add seeding for startswith and endswith

Priapos1004 committed:
Fixed some of the ruff errors

- fixed version of ruff in GitHub workflow

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/strategy_selector.py:
```diff
@@ -5,6 +5,7 @@
 from pynguin.custom_seeding.schema.main_seeder_schema import MainSeederFunctionOutput
 from pynguin.custom_seeding.strategy.base_strategy import BaseStrategy
 from pynguin.custom_seeding.strategy.simple_strategy import SimpleStrategy
+from pynguin.custom_seeding.strategy.test_strategy import TestStrategy
```
For our final version, we should pick a better name than "test"

Priapos1004 replied:
Also than rename in `scripts/config.py` and the class itself

Priapos1004 replied:
fixed

Priapos1004 commented on pyproject.toml:
```diff
@@ -5,3 +5,18 @@ exclude = [
+[tool.uv.sources]
+pynguin = {path = "pynguin_0_41_0", editable = true }
+
+[project]
```
Why do the requirements differ from requirements.txt? This should not be the case.

Priapos1004 commented on scripts/examples/len_func/len_func_between.py:
```diff
@@ -0,0 +1,5 @@
+"""Tests behaviour of len-functions with parameter length strictly between 6 and 10."""
```
Doc-String does not match function content (10 != 12)

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
Currently, test cases can be outputted twice. Please remove duplicates in the end, e.g.

2025-06-30 14:31:56,884 | INFO | __main__ | ▶ Running Pynguin on module: str_with.startswith
2025-06-30 14:31:56,894 | INFO | pynguin.custom_seeding.strategy.test_strategy | Final Tests: [['ABCtest'], ['ABCtest']]
2025-06-30 14:31:56,894 | INFO | pynguin.custom_seeding.main_seeder | Generated 2 seeds for scripts.examples.str_with.startswith with strategy 'test'
2025-06-30 14:31:56,894 | INFO | pynguin.ga.generationalgorithmfactory | Parsed testcases: 2
2025-06-30 14:31:56,913 | INFO | pynguin.generator | Used search time: 0/10

Lexolef replied:
Currently this is expected to happen for modified parameters. It will be changed in the final version.

Priapos1004 commented on scripts/examples/len_func/len_func_perform1.py:
```diff
@@ -0,0 +1,11 @@
+"""Tests behaviour of len-functions where the longer of two parameters has odd length and lengths are unequal."""
+def len_func_perform1(param1: str, param2: str) -> bool:
+    len1 = len(param1)
```
I think this test is not in the right format, as we are not checking if-statements of variables other than input parameter. (no seeds are generated as expected). Correction would be to write the len-function calls into the if-statements even if this is redundant

K92hoang replied:
all performance tests updated

Priapos1004 commented on scripts/examples/len_func/len_func_perform2.py:
```diff
@@ -0,0 +1,8 @@
+"""Tests behaviour of len-functions with two parameters with a for loop."""
+def len_func_perform2(param1: str, param2: str) -> bool:
+    len1 = len(param1)
```
I think this test is not in the right format, as we are not checking if-statements of variables other than input parameter. (no seeds are generated as expected). Correction would be to write the len-function calls into the if-statements even if this is redundant

Priapos1004 commented on scripts/examples/len_func/len_func_perform3.py:
```diff
@@ -0,0 +1,9 @@
+"""Tests behaviour of len-functions with two parameters with a while loop."""
+def len_func_perform3(param1: str, param2: str) -> bool:
+    len1 = len(param1)
```
I think this test is not in the right format, as we are not checking if-statements of variables other than input parameter. (no seeds are generated as expected). Correction would be to write the len-function calls into the if-statements even if this is redundant

Priapos1004 commented on scripts/examples/len_func/len_func_perform4.py:
```diff
@@ -0,0 +1,8 @@
+"""Tests behaviour of len-functions with three parameters in three nested if statements."""
+def len_func_perform4(param1: str, param2: str, param3: str) -> bool:
+    avg = (len(param1) + len(param2) + len(param3)) / 3
```
I think this test is not in the right format, as we are not checking if-statements of variables other than input parameter. (no seeds are generated as expected). Correction would be to write the len-function calls into the if-statements even if this is redundant

Priapos1004 commented on scripts/examples/len_func/len_func_perform5.py:
```diff
@@ -0,0 +1,14 @@
+"""Tests behaviour of len-functions with three parameters in six nested if statements."""
+def len_func_perform5(param1: str, param2: str, param3: str) -> bool:
+    len12 = len(param1) + len(param2) 
```
I think this test is not in the right format, as we are not checking if-statements of variables other than input parameter. (no seeds are generated as expected). Correction would be to write the len-function calls into the if-statements even if this is redundant

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+
+    @staticmethod
+    def call_list_visit(children: list[NodeNG], param_names: list[str], tests: list[list[str]],
+                        operations: tuple[str], results: list[list[str]] | None = None) -> None:
```
You can specify that the func returns "None" with "-> None"; however, usually, if the function returns "None", the type hint is omitted, e.g.

def call_list_visit(children: list[NodeNG], param_names: list[str], tests: list[list[str]], operations: tuple[str], results: list[list[str]] | None = None):

Lexolef replied:
Thank you! I changed this.

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
Comments should always start capitalised because this is the style of pynguin, i.e. 
"# [A-Z]..."
;-)

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                    and isinstance(elt.value, str)]
+        if isinstance(node, Name):
+            # when a variable is checked, it is returned here
+            # Currently untested how this interacts or what is returned
```
How can we test this case? Is this relevant for us?

Lexolef replied:
This would include amongst others param, param comparisons. I was not sure if we want to include some basic comparisons in the future and left it in.

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+        # logger warning when no parseable statement is found
+        # should the program stop here or return nothing?
+        logger.warning("%s could not be parsed.", str(node))
+        return None
```
Either you change the output type hint to "list[str] | None" or cleaner would be to return an empty list and handle the case of an empty list in a higher layer (where this func is called).

Lexolef replied:
Changed it to empty list. Now it will not create an error upon encountering this case.

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+        return None
+
+    @staticmethod
+    def append_results(param_name: str, values: list[str], results: list[list[str]]) -> None:
```
This feels a bit unclean to modify the input parameter. Cleaner would be to return a new "list[list[str]]" and call the func like

results += append_results(...)

to add the new values. Especially, because you iterate over an object with a for-loop that is changed while doing. In other programming languages, this could lead to an error (lucky you that we work in python XD) - I would still recommend to fix this

Priapos1004 replied:
fixed

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                    logger.debug("Sorted value %s into the result list %s", val, result)
+                    found = True
+                    break
+            if not found:
```
Am I correct to assume that the "found" variable logic is to avoid duplicates?

If so, then there is a semantic error (see upper comment about duplicates)

Lexolef replied:
This is not the case. It is there to check whether a parameter already has conditions or not. 😄

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                    left = node.left
+                    right = comparator
+
+                    # Testcase: param is in something
```
Shouldn't it be:

\# Testcase: param in something

instead of

\# Testcase: param is in something

?

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+        Checks a node for specific operations (if with 'or' or 'and').
+        It then traverses the tree in a way that the results are correctly build to
+        guarantee the right way to build the testcase inputs following the extraction.
+        ATTENTION!!! While and for loops create no wrong tests,
```
Like mentioned above, there should be no duplicates in the tests handed-back to main_seeder.py

Lexolef replied:
For and While loops are in this version not handled. Therefore it is currently out of our test scope. We will resolve this in the future.

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+        ATTENTION!!! While and for loops create no wrong tests,
+        but multiples of the same testcase.
+        """
+        results = [] if results is None else [result.copy() for result in results]
```
results.copy() only creates shallow-copies, i.e. hands over the reference and in changing the copy, the original will be changed. If you want deepcopies of the list[str] in results, use:

import copy
copy.deepcopy(\<list object\>)

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                        and isinstance(left, Const)
+                        ):
+                        values = TestStrategy._extract_literal_repr(left)
+                        TestStrategy.append_results(right.name, values, results)
```
This function should also hand-down the found parameters and not work with by-reference changes, as this is the cleaner way and can prevent nasty bugs because of business logic in unexpected layers (keep it simple and robust).

Priapos1004 replied:
Fixed

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+        if isinstance(node, If):
+            TestStrategy.find_parameters(node.test, param_names, results, operations)
+
+            # checks for 'and' comparisons and accumulates results over them
```
What happens in the case:

if ... and ... or ...:

I think we should not handle this for now, but the question is if this is currently skipped or leads to an error.

Lexolef replied:
It will generate a test only for the outer compare. That means in case of P1 and P2 or P3 he will only check P1. Same for or in the first statement. With paranthesis () he will only check the outer one. I think this is fine now, since he will generate a testcase based on the statements within that if check as well.

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                    TestStrategy.call_list_visit(node.body, param_names, tests, operations, results)
+                    results = current_results
+
+            if (isinstance(node.test, Call)
```
For readability, I would recommend to have the brackets in separate lines:
```text
if (
       isinstance(node.test, Call)
       and isinstance(node.test.func, Attribute)
       and node.test.func.attrname in {"startswith", "endswith"}
):
```

Lexolef replied:
Outdated

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                    TestStrategy.find_parameters(condition, param_names, results, operations)
+
+            # checks for 'or' comparisons and accumulates different results over the possibilities
+            if isinstance(node.test, BoolOp) and node.test.op == "or":
```
If there are multiple if-statements that are disjunctive, use "if-elif-else" instead of multiple if-statements and one else-statement

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+
+            # checks for 'or' comparisons and accumulates different results over the possibilities
+            if isinstance(node.test, BoolOp) and node.test.op == "or":
+                current_results = results.copy()
```
Again, if this is intented to be a deepcopy use the "copy" library

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                        idx = param_names.index(result[0])
+                        testcase[idx] = "".join(result[1:])
+                tests.append(testcase)
+                logger.debug("Appended test from startswith/endswith: %s", testcase)
```
To much detailed logic for visit function. The depth of logic should be the same in one function - I recommend to create a new function for this

Priapos1004 replied:
fixed

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                TestStrategy.call_list_visit(node.body, param_names, tests, operations, results)
+        # calls children of any other node
+        else:
+            TestStrategy.call_list_visit(node.get_children(), param_names, tests, operations,
```
Technically, "node.get_children()" is of type "Iterator[NodeNG]" so you should pass on "list(node.get_children())" instead (for our project size not relevant, but in bigger projects such minor type conflicts can result in hard-to-find bugs)

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                        TestStrategy.append_results(right.name, values, results)
+
+    @staticmethod
+    def visit(node: NodeNG, param_names: list[str], tests: list[list[str]], operations: tuple[str],
```
If possible, do not use by-reference changes if you are using py lists and the data amount is small, as lists are linked-lists and therefore, using append or creating a new list and adding it to the old outside the function should not make a difference -> but is easier to debug and more robust against changes.

Priapos1004 replied:
fixed

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                                         results)
+        # extracts testcases from results
+        if TestStrategy.is_leaf_node(node) and results:
+            logger.debug("%s", node.lineno)
+            for result in results:
+                logger.debug("Parsed results in visit: %s", results)
+                idx = param_names.index(result[0])
+                testcase[idx] = "".join(result[1:])
+
+            tests.append(testcase)
+            logger.debug("Found visit test: %s", tests)
```
This could be out-sourced to have the visit function as some kind of "controller" which manages function calls, but has no technical logic inside. Would be a bit cleaner and ruff will be happier :-)

Priapos1004 replied:
In general, the whole starts_with logic is redundant and introduces a lot of technical debt. It should be integrated in a way that keeps the visit-func simple -> the visit-func should only handle the abstract logic of traversing and handling "or" and "and". Everthing function (in, len, starts_with, ...) related should be implemented elsewhere.

Priapos1004 replied:
fixed

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+            tests.append(testcase)
+            logger.debug("Found visit test: %s", tests)
+
+    def _extract_in_comparisons(self, ast_tree: FunctionDef | AsyncFunctionDef) -> list[list[str]]:
+        """Finds 'in' and 'not in' statements of input parameters.
+
+        Calls the visit for all parts of the given ast tree,
+        checking for 'in' and 'not in' parameters.
+        At the end the list of testcase lists is returned.
+        SUGGESTION!!! Based on how our program can handle other comparisons
+        we could scrap this function in the future and call from _generate_test_cases directly.
+        """
+        tests = []
+        input_parameters = self.get_parameter_names()
+        for statement in ast_tree.body:
+            TestStrategy.visit(statement, input_parameters, tests, ("in", "not in"))
+        return tests
+
+    def _extract_startswith_endswith(self) -> list[list[list]]:
+        """Finds startswith and endswith statements of input parameters."""
+        tests = []
+        input_parameters = self.get_parameter_names()
+        for statement in self.function_info.ast_tree.body:
+            TestStrategy.visit(statement, input_parameters, tests, ("startswith", "endswith"))
+        return tests
```
redundant...

Priapos1004 reviewed (changes requested):
There are multiple things that need changing:

- The test functions have little things that need changing (added some small comments)
- It would be cool to have test-cases for 'if "..." in ["...", "....", ...]:' (maybe one plain with two strings in list, one with 5 strings in list, and one nested with multiple if-statements some with lists and some with substring comparison)
- The general idea of the TestStrategy (pls change name XD) is good. However, the current way of mixing logic levels is not very modular and introduces technical debt. Especially, the current state of the "visit" function is not good. Here, I recommend to separate more strictly between traversing, handling specific operators etc. (see comments)

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+        # logger warning when no parseable statement is found
+        # should the program stop here or return nothing?
+        logger.warning("%s could not be parsed.", str(node))
+        return None
```
I also suggested this XD

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+            TestStrategy.find_parameters(node.test, param_names, results, operations)
+
+            # checks for 'and' comparisons and accumulates results over them
+            if isinstance(node.test, BoolOp) and node.test.op == "and":
+                for condition in node.test.values:
+                    TestStrategy.find_parameters(condition, param_names, results, operations)
+
+            # checks for 'or' comparisons and accumulates different results over the possibilities
+            if isinstance(node.test, BoolOp) and node.test.op == "or":
```
@Lexolef Is this relevant/true?

Lexolef replied:
Tried it and does not work.

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
```diff
@@ -0,0 +1,226 @@
+                    TestStrategy.find_parameters(condition, param_names, results, operations)
+
+            # checks for 'or' comparisons and accumulates different results over the possibilities
+            if isinstance(node.test, BoolOp) and node.test.op == "or":
```
Same here?

Lexolef replied:
Does not work here as well.

Priapos1004 commented on scripts/examples/len_func/len_func_between.py:
```diff
@@ -0,0 +1,5 @@
+"""Tests behaviour of len-functions with parameter length strictly between 6 and 10."""
+def len_func_between(param: str) -> bool:
+    if 6 < len(param) < 12:
```
I suggested that as well XD

K92hoang replied:
docstring corrected

Priapos1004 commented on pynguin_0_41_0/src/pynguin/testcase/execution.py:
```diff
@@ -1007,7 +1007,7 @@ def log_thread_exception(arg):
             _LOGGER.error(
                 "Exception in Thread: %s",
                 arg.thread,
-                exc_info=(arg.exc_type, arg.exc_value, arg.exc_traceback),  # noqa: LOG014
+                exc_info=(arg.exc_type, arg.exc_value, arg.exc_traceback),
```
In ruff 0.12.1 this will not be triggered anymore

Priapos1004 commented on scripts/examples/len_func/len__func_param2_gt_param1.py:
Why two underscores?

K92hoang replied:
unintentional mistake, I have renamed it

Priapos1004 commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/test_strategy.py:
The "in_func_not_param" test case is currently not handled. Is this intended?

Priapos1004 replied:
Added as TODO comment

K92hoang committed:
Fixed len_func tests

K92hoang committed:
Added new in_func test with lists

K92hoang commented:
> * It would be cool to have test-cases for 'if "..." in ["...", "....", ...]:' (maybe one plain with two strings in list, one with 5 strings in list, and one nested with multiple if-statements some with lists and some with substring comparison)

I added three new in-func tests

Priapos1004 committed:
Merge pull request #7 from Priapos1004/len-func-tests

Len func tests

Priapos1004 reviewed (commented):
Test case fixes looking good

Priapos1004 committed:
Reverted changes for start_with and end_with func

Lexolef committed:
Changed none returns

Lexolef committed:
Removed unwanted None return

Lexolef committed:
Updated comments

Lexolef committed:
Capitalized comments

Lexolef committed:
Added cleaner copy

Lexolef committed:
Changed parameter order to unify over visit and find_parameters

Lexolef committed:
Added fix for elif mistake

Lexolef committed:
Added correct type cast for call_list_visit pass

Lexolef committed:
Corrected comments for ruff check

Lexolef committed:
Merge branch 'seeding_prototype' of https://github.com/Priapos1004/penguin_and_fuzzingbook into seeding_prototype

Priapos1004 committed:
Name fix

Priapos1004 committed:
Refactored TestStrategy -> TreeTraverseStrategy

- little change to parameter logic in base_strategy.py

Priapos1004 committed:
Fixed ruff errors

Priapos1004 committed:
Added additional test cases

- Test cases for duplicates as orientation for post-processing implementation in TreeTraverseStrategy
- Nested if-elif-else statements to check capabilities of Strategy
- Nested or-statement to see if Strategy handles it correctly

Priapos1004 committed:
Update tree_traverse_strategy.py

Priapos1004 committed:
Removed unnecessary code

Priapos1004 committed:
Fixed inconsistency

- GitHub Workflow uses ruff 0.12.1 and requirements.txt installs ruff 0.11.13

Priapos1004 committed:
Little fixes

- Improved comments in TreeTraverseStrategy
- Cleaned pyproject.toml

K92hoang reviewed (approved):
Good Structure
the code is well separated with clear of concerns between different methods.
Well done!
